### DISCUSSION TITLE
Small objects dataset - How to train?

### DISCUSSION THREAD

**Discussion by Andreluizfc** [ultralytics/yolov5#3470] [Q&A]
Hello!

I have the task to detect people on images of size 1920×1080. I have two problems though.

1. The problem is people are very small on the video feed, variating around 35x55 -10x20 pixels.
2. I don't have enough images to train. The camera is placed in an isolated region where very few people go.

I have found a dataset containing small images of people around 34x92 pixels.
Is there a way to train on this dataset I have found? Or should I train with big images with the whole scene containing people?

I'd appreciate any enlightenment on how to proceed with this task.

Thank you 🙏

**Comment by glenn-jocher** [comment]
👋 Hello @Andreluizfc! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Reply by Andreluizfc** [reply]
Thank you @glenn-jocher. I've read all the documentation needed for training.

Regarding the lack of data to train, what would you suggest doing?

1. Find "similar" images containing small people.
2. Fabricate new data where I "cut" people from one background and "paste" in the desired one.

In your experience, would the first case perform well? I guess the second one takes a lot of time to implement.

Cheers.

**Reply by glenn-jocher** [reply]
@Andreluizfc well in my opinion training a person detector is not in your best interest, I would simply use one of the official models instead and filter only class 0 (person) results. You can experiment with various inference settings like image size until you get your desired result. No point in reinventing the wheel.
```bash
python detect.py --classes 0
```

**Reply by Andreluizfc** [reply]
> @Andreluizfc well in my opinion training a person detector is not in your best interest, I would simply use one of the official models instead and filter only class 0 (person) results. You can experiment with various inference settings like image size until you get your desired result. No point in reinventing the wheel.
> 
> ```shell
> python detect.py --classes 0
> ```

Dear @glenn-jocher 

I tried running predictions as you said  `python detect.py --classes 0` plus using the `--img 1280` param. I also tried `--img 1920`.
The results were better, but at some frames, the person is too far to detect.

I was reading the following [article](https://www.robots.ox.ac.uk/~vgg/publications/2018/Neumann18a/neumann18a.pdf) to get some insights. The authors "_...trained it on the MS COCO training set, where every image was down-sampled by a factor of 4_".

Let's say I use downsampled images from [Citypersons](https://paperswithcode.com/dataset/citypersons) dataset or even MS COCO and use them to fine-tune a yolov5 model. My guess is we can achieve good results, right?

Thanks a lot!

---
References:
[Scale Match for Tiny Person Detection](https://arxiv.org/pdf/1912.10664.pdf)
[Tiny People Pose](https://www.robots.ox.ac.uk/~vgg/publications/2018/Neumann18a/neumann18a.pdf)

**Reply by glenn-jocher** [reply]
@Andreluizfc my example code is simply to show how to use the classes argument, it's not meant to be an explicit solution to your problem. As I said before, you should experiment with your inference settings, including models obviously. The default command will use YOLOv5s, the worst performant model of all 8 YOLOv5 models. See release notes for full details.
https://github.com/ultralytics/yolov5/releases/tag/v5.0

**Reply by shang0085** [reply]
Any one tried VisDrone dataset? Same problem with small objects. 
On default 10 classes, hyper parameters at scratch, img-size at 640, yolov5s pretrained weights

Results on Test set

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/soonheng/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/soonheng/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | P | R | mAP 0.5 | mAP 0.95
-- | -- | -- | -- | --
all | 0.42 | 0.313 | 0.277 | 0.147
pedestrian | 0.42 | 0.313 | 0.277 | 0.147
people | 0.45 | 0.259 | 0.235 | 0.087
bicycle | 0.422 | 0.164 | 0.141 | 0.0422
car | 0.603 | 0.728 | 0.686 | 0.411
van | 0.39 | 0.36 | 0.316 | 0.194
truck | 0.392 | 0.368 | 0.306 | 0.174
tricycle | 0.253 | 0.211 | 0.126 | 0.0599
awning-tricyle | 0.391 | 0.156 | 0.144 | 0.0752
bus | 0.635 | 0.496 | 0.51 | 0.327
motor | 0.415 | 0.303 | 0.234 | 0.08



</body>

</html>


I further went down to skipped human, bicycle classes and came down to 6 classes
img-size trained at 960. Freeze first 10 layers, yolov5l pretrained weights. Results on test set

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/soonheng/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/soonheng/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | P | R | mAP 0.5 | mAP 0.95
-- | -- | -- | -- | --
all | 0.552 | 0.454 | 0.452 | 0.281
car | 0.726 | 0.768 | 0.764 | 0.486
van | 0.483 | 0.412 | 0.407 | 0.271
truck | 0.509 | 0.45 | 0.435 | 0.279
awning-tricyle | 0.374 | 0.174 | 0.161 | 0.0926
bus | 0.707 | 0.554 | 0.604 | 0.425
motor | 0.512 | 0.367 | 0.34 | 0.134



</body>

</html>

@glenn-jocher do you want to share your hyper parameters file used for VisDrone dataset?

**Reply by glenn-jocher** [reply]
@shang0085 for small objects you need to train at larger image sizes, i.e. 1280, 2560 etc. A smaller model at larger image sizes, i.e. YOLOv5m6 will probably be better than a large model at small image sizes.

We don't have VisDrone-specific hyps. All hyps are located in data/hyps:
https://github.com/ultralytics/yolov5/tree/master/data/hyps
